File: iroha-android/src/main/java/io/soramitsu/irohaandroid/qr/QRCodeGenerator.java
package io.soramitsu.irohaandroid.qr;

import android.graphics.Bitmap;
import android.graphics.Color;

import com.google.zxing.EncodeHintType;
import com.google.zxing.WriterException;
import com.google.zxing.qrcode.decoder.ErrorCorrectionLevel;
import com.google.zxing.qrcode.encoder.ByteMatrix;
import com.google.zxing.qrcode.encoder.Encoder;
import com.google.zxing.qrcode.encoder.QRCode;

import java.util.HashMap;
import java.util.Map;

public class QRCodeGenerator {
    public static final String ENCODE_CHARACTER_TYPE_ISO_8859_1 = "ISO-8859-1";
    public static final String ENCODE_CHARACTER_TYPE_UTF_8 = "UTF-8";

    private QRCodeGenerator() {
    }

    public static Bitmap generateQR(String qrText, int size, String encodeCharacterSet) throws WriterException {
        Map<EncodeHintType,String> hints = new HashMap<>();
        hints.put(EncodeHintType.CHARACTER_SET, encodeCharacterSet);
        QRCode qrCode = Encoder.encode(qrText, ErrorCorrectionLevel.H, hints);
        ByteMatrix byteMatrix = qrCode.getMatrix();
        Bitmap bitmap = Bitmap.createBitmap(byteMatrix.getWidth(), byteMatrix.getHeight(), Bitmap.Config.ARGB_8888);
        for (int y = 0; y < byteMatrix.getHeight(); ++y) {
            for (int x = 0; x < byteMatrix.getWidth(); ++x) {
                byte val = byteMatrix.get(x, y);
                bitmap.setPixel(x, y, val == 1 ? Color.BLACK : Color.WHITE);
            }
        }
        bitmap = Bitmap.createScaledBitmap(bitmap, size, size, false);
        return bitmap;
    }
}

File: app/src/main/java/jp/co/soramitsu/irohaandroid/util/QrCodeGenerator.kt
package jp.co.soramitsu.irohaandroid.util

import android.graphics.Bitmap
import com.google.zxing.EncodeHintType
import com.google.zxing.qrcode.decoder.ErrorCorrectionLevel
import com.google.zxing.qrcode.encoder.Encoder

class QrCodeGenerator(
    private val firstColor: Int,
    private val secondColor: Int
) {

    companion object {
        private const val RECEIVE_QR_SCALE_SIZE = 1024
        private const val PADDING_SIZE = 2
    }

    fun generateQrBitmap(input: String): Bitmap {
        val hints = HashMap<EncodeHintType, String>()
        hints[EncodeHintType.CHARACTER_SET] = "UTF-8"
        val qrCode = Encoder.encode(input, ErrorCorrectionLevel.H, hints)
        val byteMatrix = qrCode.matrix
        val width = byteMatrix.width + PADDING_SIZE
        val height = byteMatrix.height + PADDING_SIZE
        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        for (y in 0 until height) {
            for (x in 0 until width) {
                if (y == 0 || y > byteMatrix.height || x == 0 || x > byteMatrix.width) {
                    bitmap.setPixel(x, y, secondColor)
                } else {
                    bitmap.setPixel(x, y, if (byteMatrix.get(x - PADDING_SIZE / 2, y - PADDING_SIZE / 2).toInt() == 1) firstColor else secondColor)
                }
            }
        }
        return Bitmap.createScaledBitmap(bitmap, RECEIVE_QR_SCALE_SIZE, RECEIVE_QR_SCALE_SIZE, false)
    }
}